ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/OperationsReOrgManagerTest.java

```diff
@@ -0,0 +1,139 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.statetransition;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.atLeastOnce;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.NavigableMap;
+import java.util.Optional;
+import java.util.TreeMap;
+import org.apache.tuweni.bytes.Bytes32;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+import tech.pegasys.teku.datastructures.blocks.BeaconBlock;
+import tech.pegasys.teku.datastructures.operations.Attestation;
+import tech.pegasys.teku.datastructures.operations.AttesterSlashing;
+import tech.pegasys.teku.datastructures.operations.ProposerSlashing;
+import tech.pegasys.teku.datastructures.operations.SignedVoluntaryExit;
+import tech.pegasys.teku.datastructures.util.AttestationProcessingResult;
+import tech.pegasys.teku.datastructures.util.DataStructureUtil;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.statetransition.attestation.AggregatingAttestationPool;
+import tech.pegasys.teku.statetransition.attestation.AttestationManager;
+import tech.pegasys.teku.storage.client.RecentChainData;
+
+@SuppressWarnings("unchecked")
+public class OperationsReOrgManagerTest {
+
+  private DataStructureUtil dataStructureUtil = new DataStructureUtil();
+
+  private OperationPool<ProposerSlashing> proposerSlashingOperationPool = mock(OperationPool.class);
+  private OperationPool<AttesterSlashing> attesterSlashingOperationPool = mock(OperationPool.class);
+  private OperationPool<SignedVoluntaryExit> exitOperationPool = mock(OperationPool.class);
+  private AggregatingAttestationPool attestationPool = mock(AggregatingAttestationPool.class);
+  private AttestationManager attestationManager = mock(AttestationManager.class);
+
+  private RecentChainData recentChainData = mock(RecentChainData.class);
+
+  private OperationsReOrgManager operationsReOrgManager =
+      new OperationsReOrgManager(
+          proposerSlashingOperationPool,
+          attesterSlashingOperationPool,
+          exitOperationPool,
+          attestationPool,
+          attestationManager,
+          recentChainData);
+
+  @Test
+  void shouldRequeueAndRemoveOperations() throws Exception {
```

Comment:
Probably worth having a test for when blocks are removed but not added or vice-versa (ie a reorg that only changes from an empty slot to a full one and the other way).  Should work fine but that's the most common type of reorg so important to handle it right.

Reply:
I did the reOrg from an empty slot to a full one, but I'm having a hard time understanding the reOrg from full one to empty slot since attestations vote on blocks.